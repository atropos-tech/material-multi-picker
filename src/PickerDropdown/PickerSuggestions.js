import React from "react";
import { MenuItem, withStyles } from "@material-ui/core";
import { LOADING, NOT_ENOUGH_CHARACTERS, isError, suggestionsPropType } from "../utils";
import { string, func, number, any, object } from "prop-types";
import DefaultError from "./DefaultError";
import DefaultSuggestion from "./DefaultSuggestion";
import DefaultEmptyMessage from "./DefaultEmptyMessage";
import DefaultLoadingMessage from "./DefaultLoadingMessage";
import DefaultMoreCharactersMessage from "./DefaultMoreCharactersMessage";

const HIGHLIGHT_GREY_SHADE = 300;

const styles = theme => ({
    highlighted: {
        backgroundColor: theme.palette.grey[HIGHLIGHT_GREY_SHADE]
    }
});

function PickerSuggestions(
    { suggestions, getItemProps, highlightedIndex, itemToString, inputValue, SuggestionComponent = DefaultSuggestion, ErrorComponent = DefaultError, classes }
) {
    if ( isError(suggestions) ) {
        return (<ErrorComponent error={ suggestions } inputValue={ inputValue } />);
    }
    if ( suggestions === NOT_ENOUGH_CHARACTERS ) {
        return <DefaultMoreCharactersMessage />;
    }
    if ( suggestions === LOADING ) {
        return <DefaultLoadingMessage inputValue={ inputValue } />;
    }
    if ( Array.isArray(suggestions) ) {
        if ( suggestions.length ) {
            return (
                <>
                    {
                        suggestions.map((item, index) => {
                            const itemId = itemToString(item);
                            const isHighlighted = highlightedIndex === index;

                            const menuItemProps = getItemProps({
                                index,
                                item,
                                role: "menuitem",
                                className: isHighlighted ? classes.highlighted : "",
                                style: {
                                    padding: 0,
                                    height: "auto"
                                },
                            });
                            return (
                                <MenuItem className="suggestion" key={ itemId } {...menuItemProps}>
                                    <SuggestionComponent itemId={ itemId } item={ item } isHighlighted={ isHighlighted } inputValue={ inputValue } />
                                </MenuItem>
                            );
                        })
                    }
                </>
            );
        }
        if ( inputValue.length ) {
            return <DefaultEmptyMessage inputValue={ inputValue } />;
        } else {
            return false;
        }
    }

    console.error(suggestions);
    throw new Error("should never happen!");
}

PickerSuggestions.propTypes = {
    suggestions: suggestionsPropType,
    getItemProps: func.isRequired,
    highlightedIndex: number,
    itemToString: func.isRequired,
    inputValue: string.isRequired,
    SuggestionComponent: any,
    ErrorComponent: any,
    classes: object
};

export default withStyles(styles)(PickerSuggestions);
