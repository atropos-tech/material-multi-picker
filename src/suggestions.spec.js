/* eslint-env jest */
/* eslint-disable no-magic-numbers */
/* eslint-disable max-statements */

import React from "react";
import { mount } from "enzyme";
import { act } from "react-dom/test-utils";
import { resetIdCounter } from "downshift";
import MultiPicker, { NOT_ENOUGH_CHARACTERS } from "./index";
import JssProvider from "react-jss/lib/JssProvider";
import PickerSuggestions from "./PickerDropdown/PickerSuggestions";


// workaround for non-stable classnames generated in JSS
// https://github.com/mui-org/material-ui/issues/9492#issuecomment-368205258
//
// forces deterministic classnames that may clash in a wider app (but are safe in unit tests)
//
// Should be addressed in future material-ui release, see
//   https://github.com/mui-org/material-ui/issues/14357
function mountStable(reactElement, ...other) {
    const generateClassName = (rule, styleSheet) => `${styleSheet.options.classNamePrefix}-${rule.key}`;
    const elementWithStableClassNames = (
        <JssProvider generateClassName={ generateClassName }>
            { reactElement }
        </JssProvider>
    );
    return mount(elementWithStableClassNames, ...other);
}

const NOOP = () => { /* do nothing */ };
const SHORT_DELAY_MILLISECONDS = 30;

function delay(delayInMilliseconds = SHORT_DELAY_MILLISECONDS) {
    return new Promise(resolve => setTimeout(resolve, delayInMilliseconds) );
}

const BASE_PROPS = {
    value: [],
    onChange: NOOP,
    itemToString: item => item,
    getSuggestedItems: () => []
};

async function changeInputValueAndUpdate(wrapper, newInputValue) {
    act(() => {
        wrapper.find("input").simulate("change", { target: { value: newInputValue}});
    });
    await delay(200);
    wrapper.update();
}

describe("suggestion dropdown", () => {

    // avoid non-deterministic id creation for downshift subcomponents
    // breaking snapshot testing
    beforeEach(resetIdCounter);

    it("does not show suggestion items that have already been picked", async () => {
        expect.assertions(4);

        const wrapper = mountStable(<MultiPicker {...BASE_PROPS } value={["some picked suggestion"]} getSuggestedItems={ () => ["some suggestion", "some picked suggestion"] } />);
        expect(wrapper).not.toContainMatchingElement(PickerSuggestions);

        await changeInputValueAndUpdate(wrapper, "some text");

        expect(wrapper).toContainExactlyOneMatchingElement(PickerSuggestions);
        expect(wrapper.find(PickerSuggestions)).toHaveText("some suggestion");
        expect(wrapper.find(PickerSuggestions).getDOMNode()).toMatchSnapshot();
    });

    it("delays fetching suggestions if throttle value is set", async () => {
        expect.assertions(3);

        const getSuggestedItems = jest.fn(() => ["some suggestion"]);
        const wrapper = mountStable(<MultiPicker {...BASE_PROPS } getSuggestedItems={ getSuggestedItems } fetchDelay={ 20} />);

        wrapper.find("input").simulate("change", { target: { value: "s"}});
        await delay(10);
        expect(getSuggestedItems).not.toHaveBeenCalled();

        wrapper.find("input").simulate("change", { target: { value: "so"}});
        await delay(10);
        expect(getSuggestedItems).not.toHaveBeenCalled();

        wrapper.find("input").simulate("change", { target: { value: "som"}});
        await delay(50);
        expect(getSuggestedItems).toHaveBeenCalledWith("som", []);
    });

    it("shows a loading message if the suggestions are being loaded", async () => {
        expect.assertions(3);

        const wrapper = mountStable(<MultiPicker {...BASE_PROPS } getSuggestedItems={ () => new Promise(NOOP) }/>);
        await changeInputValueAndUpdate(wrapper, "some text");

        expect(wrapper).toContainExactlyOneMatchingElement(PickerSuggestions);
        expect(wrapper.find(PickerSuggestions)).toHaveText("Loading suggestions for some textâ€¦");
        expect(wrapper.find(PickerSuggestions).getDOMNode()).toMatchSnapshot();
    });

    it("shows an error message if the getSuggestions function throws an error", async () => {
        expect.assertions(2);

        const getSuggestedItems = () => {
            throw new Error("fail");
        };
        const wrapper = mountStable(<MultiPicker {...BASE_PROPS } getSuggestedItems={ getSuggestedItems } />);
        await changeInputValueAndUpdate(wrapper, "some text");

        expect(wrapper.find(PickerSuggestions)).toContainExactlyOneMatchingElement("Typography.suggestion-error-message");
        expect(wrapper.find(PickerSuggestions).getDOMNode()).toMatchSnapshot();
    });

    it("shows an error message if the getSuggestions function returns a failed promise", async () => {
        expect.assertions(2);

        const getSuggestedItems = () => Promise.reject(new Error("fail"));
        const wrapper = mountStable(<MultiPicker {...BASE_PROPS } getSuggestedItems={ getSuggestedItems } />);
        await changeInputValueAndUpdate(wrapper, "some text");

        expect(wrapper).toContainExactlyOneMatchingElement("Typography.suggestion-error-message");
        expect(wrapper.find(PickerSuggestions).getDOMNode()).toMatchSnapshot();
    });

    it("shows a 'type more characters' message if the getSuggestedItems() function returns the special symbol", async () => {
        expect.assertions(2);

        const wrapper = mountStable(<MultiPicker {...BASE_PROPS } getSuggestedItems={ () => NOT_ENOUGH_CHARACTERS } />);
        await changeInputValueAndUpdate(wrapper, "some text");

        expect(wrapper).toContainExactlyOneMatchingElement("Typography.more-characters-message");
        expect(wrapper.find(PickerSuggestions).getDOMNode()).toMatchSnapshot();
    });

    it("shows a 'no suggestions found' message if the getSuggestions function returns an empty array", async () => {
        expect.assertions(2);

        const wrapper = mountStable(<MultiPicker {...BASE_PROPS }/>);
        await changeInputValueAndUpdate(wrapper, "some text");

        expect(wrapper).toContainExactlyOneMatchingElement("Typography.no-suggestions-message");
        expect(wrapper.find(PickerSuggestions).getDOMNode()).toMatchSnapshot();
    });

    it("does not make an initial call to 'getSuggestedItems' on mount", async () => {
        expect.assertions(1);

        const getSuggestedItems = jest.fn(() => []);
        mountStable(<MultiPicker {...BASE_PROPS } getSuggestedItems={ getSuggestedItems } />);

        await delay(50);

        expect(getSuggestedItems).not.toHaveBeenCalled();
    });
});
