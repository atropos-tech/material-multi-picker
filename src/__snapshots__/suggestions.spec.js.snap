// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`suggestion dropdown does not show suggestion items that have already been picked 1`] = `
<li
  aria-selected="false"
  class="MuiButtonBase-root MuiListItem-root MuiListItem-default MuiListItem-gutters MuiListItem-button MuiMenuItem-root MuiMenuItem-gutters"
  id="downshift-0-item-0"
  role="menuitem"
  style="padding: 0px; height: auto;"
  tabindex="-1"
>
  <p
    class="MuiTypography-root MuiTypography-body1"
    style="padding: 11px 16px;"
  >
    some suggestion
  </p>
  <span
    class="MuiTouchRipple-root"
  />
</li>
`;

exports[`suggestion dropdown shows a 'no suggestions found' message if the getSuggestions function returns an empty array 1`] = `
<h6
  class="MuiTypography-root MuiTypography-subtitle1 MuiTypography-alignCenter no-suggestions-message"
>
  No suggestions found for 
  <strong>
    some text
  </strong>
</h6>
`;

exports[`suggestion dropdown shows a 'type more characters' message if the getSuggestedItems() function returns the special symbol 1`] = `
<h6
  class="MuiTypography-root MuiTypography-subtitle1 MuiTypography-alignCenter more-characters-message"
>
  Type more characters to see suggestions
</h6>
`;

exports[`suggestion dropdown shows a loading message if the suggestions are being loaded 1`] = `
<h6
  class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom MuiTypography-alignCenter"
>
  Loading suggestions for 
  <strong>
    some text
  </strong>
  â€¦
</h6>
`;

exports[`suggestion dropdown shows an error message if the getSuggestions function returns a failed promise 1`] = `
<h6
  class="MuiTypography-root MuiTypography-h6 MuiTypography-alignCenter DefaultError-root suggestion-error-message"
>
  An error occurred!
</h6>
`;

exports[`suggestion dropdown shows an error message if the getSuggestions function throws an error 1`] = `
<h6
  class="MuiTypography-root MuiTypography-h6 MuiTypography-alignCenter DefaultError-root suggestion-error-message"
>
  An error occurred!
</h6>
`;
